sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -667,7 +667,14 @@ object JdbcUtils extends Logging {
     val getConnection: () => Connection = createConnectionFactory(options)
     val batchSize = options.batchSize
     val isolationLevel = options.isolationLevel
-    df.foreachPartition(iterator => savePartition(
+    val numPartitions = options.numPartitions
+    val repartitionedDF =
+      if (numPartitions != null && numPartitions.toInt != df.rdd.getNumPartitions) {
```

Comment:
Increasing the number of partitions can improve the insert performance in some scenarios, I think. However, `repartition` is not cheap.